src/object-store/src/layers/lru_cache/read_cache.rs

```diff
@@ -0,0 +1,267 @@
+enum ReadResult {
+    // Read success with size
+    Success(u32),
+    // File not found
+    NotFound,
+}
+
+impl ReadResult {
+    fn size_bytes(&self) -> u32 {
+        match self {
+            ReadResult::NotFound => 0,
+            ReadResult::Success(size) => *size,
+        }
+    }
+}
+
+/// Generate an unique cache key for the read path and range.
+fn read_cache_key(path: &str, args: &OpRead) -> String {
+    format!(
+        "{:x}.cache-{}",
+        md5::compute(path),
+        args.range().to_header()
+    )
+}
+
+/// Local read cache for files in object storage
+#[derive(Clone, Debug)]
+pub(crate) struct ReadCache<C: Clone> {
+    /// Local file cache backend
+    file_cache: Arc<C>,
+    /// Local memory cache to track local cache files
+    mem_cache: Cache<String, ReadResult>,
+}
+
+impl<C: Accessor + Clone> ReadCache<C> {
+    /// Create a `[ReadCache]` with capacity in bytes.
+    pub(crate) fn new(file_cache: Arc<C>, capacity: usize) -> Self {
+        let file_cache_cloned = file_cache.clone();
+        let eviction_listener =
+            move |read_key: Arc<String>, read_result: ReadResult, _cause| -> ListenerFuture {
+                // Delete the file from local file cache when it's purged from mem_cache.
+                decrement_gauge!(OBJECT_STORE_LRU_CACHE_ENTRIES, 1.0);
+                let file_cache_cloned = file_cache_cloned.clone();
+
+                async move {
+                    if let ReadResult::Success(size) = read_result {
+                        decrement_gauge!(OBJECT_STORE_LRU_CACHE_BYTES, size as f64);
+
+                        let _ = file_cache_cloned.delete(&read_key, OpDelete::new()).await;
+                        debug!("Deleted local cache file `{}`.", read_key);
+                    }
+                }
+                .boxed()
+            };
+
+        Self {
+            file_cache,
+            mem_cache: Cache::builder()
+                .max_capacity(capacity as u64)
+                .weigher(|_key, value: &ReadResult| -> u32 {
+                    // TODO(dennis): add key's length to weight?
+                    value.size_bytes()
+                })
+                .async_eviction_listener(eviction_listener)
+                .support_invalidation_closures()
+                .build(),
+        }
+    }
+
+    /// Returns the cache's entry count and total approximate entry size in bytes.
+    pub(crate) async fn stat(&self) -> (u64, u64) {
+        self.mem_cache.run_pending_tasks().await;
+
+        (self.mem_cache.entry_count(), self.mem_cache.weighted_size())
+    }
+
+    /// Invalidte all cache items which key starts with `prefix`.
+    pub(crate) async fn invalidate_entries_with_prefix(&self, prefix: String) {
+        // Safety: always ok when building cache with `support_invalidation_closures`.
+        self.mem_cache
+            .invalidate_entries_if(move |k: &String, &_v| k.starts_with(&prefix))
+            .ok();
+        self.mem_cache.run_pending_tasks().await;
```

Comment:
question: It is necessary to run pending tasks every time we invalidate an entry or write an entry? I think we can tolerate inaccurate statistics.

Reply:
The pending tasks not only contain the tasks to update statistics but also the eviction tasks.

Reply:
I'm not sure if it hurts the performance if we call it in `read()` and `invalidate_entries()` manually since the moka cache itself can do the eviction task automatically. Though we can call it in some tests.